THRIFT-4329: Implement multiplexed processor that matches CPP and Java. Tested aga…

gadLinux opened the pull request:
Implement multiplexed processor that matches CPP and Java. Tested against Java
    Missing Test suite
    Implement stored message protocol that adds the possibility of reading a header of a message and then process it

I don't remember how to do cross tests from protocols. I want to test Java and C multiplexors against a C server. I did in my test units but Don't find the way here. I did it before so I will check. 

For now it's safe to merge this.

gadLinux commented:
We must also investigate the binary protocol in C since I think is leaking memory everywhere:

==4534== 2,399,952 bytes in 99,998 blocks are definitely lost in loss record 611 of 611
==4534==    at 0x4C2FB25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==4534==    by 0x5300580: g_malloc0 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.5306.0)
==4534==    by 0x4E49774: thrift_binary_protocol_read_string (thrift_binary_protocol.c:797)
==4534==    by 0x4E48E55: thrift_binary_protocol_read_message_begin (thrift_binary_protocol.c:431)
==4534==    by 0x4E46013: thrift_multiplexed_processor_process_impl (thrift_multiplexed_processor.c:60)
==4534==    by 0x4E51864: thrift_simple_server_serve (thrift_simple_server.c:58)
==4534==    by 0x108DAF: test_server_initialize (test_server.c:74)
==4534==    by 0x5321459: ??? (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.5306.0)
==4534==    by 0x532138A: ??? (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.5306.0)
==4534==    by 0x5321631: g_test_run_suite (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.5306.0)

jeking3 commented:
Recommend you open a different ticket for memory leaks in c_glib binary protocol.

gadLinux commented:
Hi @jeking3 don't worry I will fix the binary protocol memory leak. 

About the build I think this is failing because foreign problem, not by my changes. Cause it fails for decorator protocol that was already there since long time. Can you check it please? And merge it?

gadLinux commented:
@jeking3 a memory leak is getting me crazy maybe we can check together. In the multiplexed protocol, line 169 we should release that protocol but if I do it the code seems to fail with memory dump in other point. I think the chained destructor is making me little crazy. 

I will still check but for now we should get this out.

gadLinux commented:
@jeking3 Can you merge this, please?

jeking3 commented:
Now that the build issues seem to be out of the way, I've merged the previous 3 PRs from you and I'd ask that you squash and rebase this on the current master and submit it again.  I'd like to see a clean CI build - this one failed in more places than can be explained by the previous errors.

gadLinux commented:
Ok. I will do it.

gadLinux commented:
rebased

jeking3 commented:
Doesn't look like a rebase, unfortunately.
What I would suggest is that you refresh your master from upstream, branch off it, cherry-pick the original commits you made in order, then squash them to a single commit, and then submit a new PR.

The build failed here in job 8 with:
```
lib/libthrift_c_glib.so.1.0.0: undefined reference to `thrift_protocol_decorator_get_type'
```

jeking3 commented:
Try this:

```
git checkout master
git pull upstream master
git checkout -b THRIFT-4329-take-2
git cherry-pick 1d1d9d8
git cherry-pick 0aedc8a
git cherry-pick 3df26ec
git cherry-pick 48ee1f5
git rebase -i HEAD~4
(change lines 2 through 4 from "pick" to "fixup")
git push -u origin THRIFT-4329-take-2:THRIFT-4329 --force
```

The last command will force your local branch THRIFT-4329-take-2 into the github branch THRIFT-4329, which is your pull request.

gadLinux commented:
I just went back in history and pushed again after rebase. Now it should be ok

jeking3 commented:
Build job 4 failed in ruby testing.  Looking for a match in PRs or Jira.

jeking3 commented:
Ruby error in build job 4 is https://issues.apache.org/jira/browse/THRIFT-2913.  c_glib unit tests ran so we can overlook this failure.

gadLinux commented:
It seems the protocol decorator was removed, I suppose I copied and pasted wrong...
I always forget that the build system is Cmake and forget to test it against it.

gadLinux commented:
Perfect, now fails on things not related to my changes.

jeking3 commented:
@gadLinux I am running the cross test suite locally.  The errors in the build appear to be in c_glib...

jeking3 commented:
This passed for me locally on a full cross, so I'll merge it.

jeking3 commented:
Hmm, I ran cross tests locally and multi isn't being exercised for c_glib server:
```
root@ecb350ad7c01:/thrift/src# test/test.py --server c_glib --client java
Apache Thrift - Integration Test Suite
Sat Sep 23 00:08:59 2017
===============================================================================
server-client:          protocol:         transport:               result:
c_glib-java             binary            framed-fastframed-ip     success
c_glib-java             binary            buffered-ip              success
c_glib-java             compact           framed-fastframed-ip     success
c_glib-java             compact           buffered-ip              success
c_glib-java             compact           framed-ip                success
c_glib-java             binary            framed-ip                success
===============================================================================
```

gadLinux commented:
Now seems ok. This is another big change. So I will try to do cross tests after. Meanwhile I cross validated with Java. Already in production with this stuff...

gadLinux commented:
What's multic and multi. What's the difference?

jeking3 commented:
multic is for compact and multi is for binary

they are backwards compatible so one can use a binary client against a
multi server (which is why you will see servers with "binary:multi" or
"multi:binary" (I forget the polarity).

On Sat, Sep 23, 2017 at 11:52 AM, Gonzalo Aguilar Delgado <
notifications@github.com> wrote:

> What's multic and multi. What's the difference?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/thrift/pull/1361#issuecomment-331644242>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALOdbSgVoRydNqTy35al06i_t0MrHg3yks5slSk1gaJpZM4PYyTs>
> .
>

gadLinux commented:
@jeking3 I don't think binary and multi server can be compatible. In fact I see it failing everytime and this is why I didn't set your protocol recommendations. 

Let me explain. When a multiX server is set, the processor allows you to register several processors. Each of this processors can implement different services. If you are contacting with a binary client (not multi) when you send the message it's not prefixed by the service name. It means that at the server end the server will never know what's the destination processor for that request. 

So when you use a binary client, you must use a binary server. And with the multi happens the same.

Apart from this, the protocol is implemented, my problem is that I was confused with the protocols in the json file and I don't know what should I really enable...


Besides this I ran it manually against binary multi client and ran ok. 

test_client:test_client:93: libtool wrapper (GNU libtool) 2.4.6 Debian-2.4.6-1
test_client:test_client:114: newargv[0]: /home/gaguilar/thrift/thrift/test/c_glib/.libs/test_client
test_client:test_client:104: newargv[1]: --protocol=multi
test_client:test_client:104: newargv[2]: --transport=buffered
test_client:test_client:104: newargv[3]: --port=39347
Connecting (buffered/binary:multi) to: ip/localhost:39347
Test #1, connect localhost:39347
testVoid() = void
testString("Test") = "Test"
testSecondServiceMultiplexSecondTestString("2nd") = "2nd"
testByte(true) = true
testByte(false) = false
testByte(1) = 1
testByte(-1) = -1
testI32(-1) = -1
testI64(-34359738368) = -34359738368
testDouble(-5.2098523) = -5.209852
testBinary(empty) = empty
testBinary([-128..127]) = {-128,-127,-126,-125,-124,-123,-122,-121,-120,-119,-118,-117,-116,-115,-114,-113,-112,-111,-110,-109,-108,-107,-106,-105,-104,-103,-102,-101,-100,-99,-98,-97,-96,-95,-94,-93,-92,-91,-90,-89,-88,-87,-86,-85,-84,-83,-82,-81,-80,-79,-78,-77,-76,-75,-74,-73,-72,-71,-70,-69,-68,-67,-66,-65,-64,-63,-62,-61,-60,-59,-58,-57,-56,-55,-54,-53,-52,-51,-50,-49,-48,-47,-46,-45,-44,-43,-42,-41,-40,-39,-38,-37,-36,-35,-34,-33,-32,-31,-30,-29,-28,-27,-26,-25,-24,-23,-22,-21,-20,-19,-18,-17,-16,-15,-14,-13,-12,-11,-10,-9,-8,-7,-6,-5,-4,-3,-2,-1,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127} OK size 256 OK
testStruct({"Zero", 1, -3, -5}) = {"Zero", 1, -3, -5}
testNest({1, {"Zero", 1, -3, -5}), 5} = {1, {"Zero", 1, -3, -5}, 5}
testMap({0 => -10, 1 => -9, 3 => -7, 2 => -8, 4 => -6}) = {0 => -10, 1 => -9, 3 => -7, 2 => -8, 4 => -6}
testStringMap({"some" => "thing", "a" => "2", "b" => "blah")} = {"some" => "thing", "a" => "2", "b" => "blah"}
testSet({0, 2, -1, -2, 1}) = {1, -1, 0, 2, -2}
testList({-2, -1, 0, 1, 2}) = {-2, -1, 0, 1, 2}
testEnum(ONE) = 1
testEnum(TWO) = 2
testEnum(THREE) = 3
testEnum(FIVE) = 5
testEnum(EIGHT) = 8
testTypedef(309858235082523) = 309858235082523
testMapMap(1) = {-4 => {-4 => -4, -3 => -3, -2 => -2, -1 => -1, }, 4 => {1 => 1, 2 => 2, 3 => 3, 4 => 4, }, }
testInsanity() = {1 => {2 => {{8 => 8, 5 => 5, }, {{"Goodbye4", 4, 4, 4}, {"Hello2", 2, 2, 2}, }}, 3 => {{8 => 8, 5 => 5, }, {{"Goodbye4", 4, 4, 4}, {"Hello2", 2, 2, 2}, }}, }, 2 => {6 => {{}, {}}, }, }
testClient.testException("Xception") =>  {1001, "Xception"}
testClient.testException("TException") =>  Caught TException
testClient.testException("success") =>  void
testClient.testMultiException("Xception", "test 1") =>  {1001, "This is an Xception"}
testClient.testMultiException("Xception2", "test 2") =>  {2002, {"This is an Xception2"}}
testClient.testMultiException("success", "test 3") =>  {{"test 3"}}
testClient.testOneway(1) =>  success - took 0.03 ms
re-test testI32(-1) = -1
Total time: 25934720 us

All tests done.
Number of failures: 1
Min time: 25934720 us
Max time: 25934720 us
Avg time: 25934720 us


So please check what protocol should be enabled and enable them. If it fails again I will check.

gadLinux commented:
What I'm thinking looking at Java implementation is that on c_glib we can set a property in the processor to define what's the default handler to be get in case of a not multi client. Let me see if I can implement it.

gadLinux commented:
Ok. Fixed a bug in the test client. And added the default protocol for the multiplexed. That way the first registered protocol becomes default. This allows you to use the multi without much hassle. But you can always override the default protocol using a property. Nice...

jeking3 commented:
Seems like it still has some issues in the cross test with c_glib.

gadLinux commented:
Yes, there are a couple of them on the wild. I will take a look ASAP. I had three of my servers broke at the same time this week and I'm was crazy until today to bring them up again. Give me a couple of days.

gadLinux commented:
@jeking3 It will fail because binary protocols doesn't use limit string parameter. Can you remove them from tests since it's not supported.

gadLinux commented:
Someone broke the build system or there's a patch that made things worse. Anyway. Check tests manually. I think it's ready to be merged. In fact what I solved in last commit was not related to my changes, anyway I fixed.

gadLinux commented:
@jeking3 Do you think it would be hard to resolve? If you need me to do something. Can you point me the right direction. Remember to remove the binary protocols from the cross tests since they don't support the limi xxx parameter.

jeking3 commented:
As far as I know builds in CI are stable.  Rebase against master.

gadLinux commented:
Rebased and lots of fixes added. Crossing fingers.
I should opened a new issue but since all bug fixing were discovered while doing this incident I took the opportunity to fix everything here.

gadLinux commented:
I finally removed the binary protocol from the list of the limits feature tests. I don't know if I did right but it seems so. This bug also contains important bugfixes for the connection management. So I hope we can merge it soon.

gadLinux commented:
@jeking3 I don't know why some compilations fails. Maybe you could check because I don't know about what travis is doing. For me seems ok.
Our tests are now ok. I added some fixes to the tests.
I see failures for some languages but I think the problem should be there since some other languages are working properly

gadLinux commented:
Please merge it or give me a reason to not do it.

jeking3 commented:
If I merge this, it will break all future builds.
Errors in c_glib start at:

https://travis-ci.org/apache/thrift/jobs/300505547 line 2990
if you run the docket ubuntu-xenial image and build inside it, you will see them

https://travis-ci.org/apache/thrift/jobs/300505547 line 2985
if you run the docket ubuntu-trusty image and build inside it, you will see them
they look quite similar to the ubuntu-xenial image issues

Suggest you build inside a docker image to verify your fixes.  Here is how I do it:

This is my `~/.bash_aliases` file:
```
# Kill all running containers.
alias dockerkillall='docker kill $(docker ps -q)'

# Delete all stopped containers.
alias dockercleanc='printf "\n>>> Deleting stopped containers\n\n" && docker rm $(docker ps -a -q)'

# Delete all untagged images.
alias dockercleani='printf "\n>>> Deleting untagged images\n\n" && docker rmi $(docker images -q -f dangling=true)'

# Delete all stopped containers and untagged images.
alias dockerclean='dockercleanc || true && dockercleani'

# Build a thrift docker image (run from top level of git repo): argument #1 is image type (ubuntu, centos, etc).
function dockerbuild
{
  docker build -t $1 build/docker/$1
}

# Run a thrift docker image: argument #1 is image type (ubuntu, centos, etc).
function dockerrun
{
  docker run -v $(pwd):/thrift/src -it $1 /bin/bash
}
```

To pull down the current image being used to build (the same way Travis CI does it) - if it is out of date in any way it will build a new one for you.
```
thrift$ DOCKER_REPO=thrift/thrift-build DISTRO=ubuntu-xenial build/docker/refresh.sh
```

To run all unit tests (just like Travis CI):
```
thrift$ dockerrun ubuntu-xenial
root@8caf56b0ce7b:/thrift/src# build/docker/scripts/autotools.sh
```

To run the cross tests (just like Travis CI):
```
thrift$ dockerrun ubuntu-xenial
root@8caf56b0ce7b:/thrift/src# build/docker/scripts/cross-test.sh
```

When you are done, you want to clean up occasionally so that docker isn't using lots of extra disk space:
```
thrift$ dockerclean
```

You need to run the docker commands from the root of the git repository for them to work.

When you are done in the root docker shell you can `exit` to go back to your user host shell.  Once the unit tests and cross test passes locally, then submit he changes, and squash the pull request to one commit to make it easier to merge.  Thanks.  I am going to update the docker README.md with this information so others can leverage it too.  Now you are building like Travis CI does!

gadLinux commented:
Unfortunately it doesn't work for me on the second command. 
dockerrun ubuntu-xenial
Unable to find image 'ubuntu-xenial:latest' locally
docker: Error response from daemon: repository ubuntu-xenial not found: does not exist or no pull access.

I'm trying to simulate the docker build in my desktop. It seems that some C99 flags are added to the build done by travis. Something is not in the normal builds. This should be resolved at code level not at travis build level. 

Anyway I will check why in travis it fails.

Thank you for pointing out this to me.

gadLinux commented:
With the new image build/docker/scripts/autotools.sh ran ok. 
I'm running the other... But no error sofar...
I will try to add -pendantic to the compilation if it doesn't have it in my host

gadLinux commented:
Ok I was unable to run as you said in travis. But the errors the code was giving was because somewhere in build system is configured to compile some binaries as C90. The errors where mostly comments and some initialization C99 style variables. Most of them outside the code I touched on this feature.

It now there's a requirement to write code with C90 style we have a problem because generators are not ready. In fact I had to disable parts of the compilation stuff to check where the problem were because the generated code is invalid in C90 style. 

I think the build system has to be updated or the rules to write code to be updated. So please if this fails again, I will let you correct because you know a lot more about the build system than I do. 

Thank you for your support.

jeking3 commented:
Try "dockerbuild ubuntu-xenial" then "dockerrun ubuntu-xenial", then "build/docker/scripts/autotools.sh" inside the docker environment as that is how travis runs, you will have the exact same build.  There are no special flags being added by Travis CI.

gadLinux commented:
@jeking3 It doesn't fail for me, even in the docker container. Please review and merge.

gadLinux commented:
4197.6 Fails... I'm done. Don't know what else to do.
Please check what is it...

jeking3 commented:
Job 6 failed due to https://issues.apache.org/jira/browse/THRIFT-2913 which pops up far too often.
Job 4 failed legitimitely and indicates there is still something that needs to be fixed:
```
===============================================================================
*** Following 24 failures were unexpected ***:
If it is introduced by you, please fix it before submitting the code.
===============================================================================
server-client:          protocol:         transport:               result:
c_glib-csharp           multi-binary      framed-ip                failure(timeout)
c_glib-csharp           multic-compact    buffered-ip              failure(timeout)
c_glib-csharp           multic-compact    framed-ip                failure(timeout)
c_glib-csharp           multi-binary      buffered-ip              failure(timeout)
c_glib-rs               multic            buffered-ip              failure(timeout)
c_glib-rs               multic            framed-ip                failure(timeout)
c_glib-rs               multi             buffered-ip              failure(timeout)
c_glib-rs               multi             framed-ip                failure(timeout)
java-c_glib             multic            buffered-ip              failure(1)
java-c_glib             multic            fastframed-framed-ip     failure(1)
java-c_glib             multic            buffered-ip-ssl          failure(1)
java-c_glib             multic            fastframed-framed-ip-ssl failure(1)
java-c_glib             multic            framed-ip                failure(1)
java-c_glib             multic            framed-ip-ssl            failure(1)
java-c_glib             multi             buffered-ip              failure(1)
java-c_glib             multi             buffered-ip-ssl          failure(1)
java-c_glib             multi             fastframed-framed-ip     failure(1)
java-c_glib             multi             fastframed-framed-ip-ssl failure(1)
java-c_glib             multi             framed-ip                failure(1)
java-c_glib             multi             framed-ip-ssl            failure(1)
rs-c_glib               multic            buffered-ip              failure(1)
rs-c_glib               multi             buffered-ip              failure(1)
rs-c_glib               multic            framed-ip                failure(1)
rs-c_glib               multi             framed-ip                failure(1)
===============================================================================
Unexpected failures are logged to test/log/unexpected_failures.log
You can browse results at:
	file:///thrift/src/test/index.html
# If you use Chrome, run:
# 	cd /thrift/src
#	python -m http.server 8001
# then browse:
# 	http://localhost:8001/test/
Full log for each test is here:
	test/log/server_client_protocol_transport_client.log
	test/log/server_client_protocol_transport_server.log
24 failed of 2590 tests in total.
Test execution took 2038.7 seconds.
Mon Nov 13 09:35:57 2017
Makefile:1097: recipe for target 'cross-.*' failed
make: *** [cross-.*] Error 1

RET=$?
if [ $RET -ne 0 ]; then
  cat test/log/unexpected_failures.log
fi
TEST: [c_glib-csharp_multi-binary_framed-ip]
*************************** server message ***************************
Mon Nov 13 09:03:46 2017
Executing: /thrift/src/test/c_glib/test_server --lt-debug --protocol=multi --transport=framed --port=36027
Directory: /thrift/src/test/c_glib
config:delay: 5
config:timeout: 9
===============================================================================
test_server:test_server:93: libtool wrapper (GNU libtool) 2.4.6 Debian-2.4.6-0.1
test_server:test_server:114: newargv[0]: /thrift/src/test/c_glib/.libs/lt-test_server
test_server:test_server:104: newargv[1]: --protocol=multi
test_server:test_server:104: newargv[2]: --transport=framed
test_server:test_server:104: newargv[3]: --port=36027
Starting "simple" server (framed/binary:multi) listen on: 36027

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: message type invalid for this processor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Success
** Message: thrift_simple_server_serve: failed to read 4 bytes - Success

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: message type invalid for this processor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor

(lt-test_server:29796): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Bad file descriptor
```

gadLinux commented:
Yes but this seems to be in other languages that are not in my scope. So it should be others that fix them. So I think it's good enough for now. Can you merge it? And maybe open a new bug to trac them?

jeking3 commented:
@gadLinux I think you need to re-read the error messages.  It says GLib-WARNING and GError.  That's the c_glib library.  Further it looks like all the "multi" protocol tests failed with c_glib on either the server or client side.  I cannot merge this - it will break all future builds.

You can run these tests in a docker sandbox, after "make precross" then you can do `test/test.py --client c_glib` to limit the tests to just c_glib clients.

gadLinux commented:
@jeking3 I don't think is an error, but that all this cross tests are directly not supported. No all protocols are implemented.

gadLinux commented:
What I'm telling you is that I need help if you want to fix everything on this language. I don't have all the time to implement the full stack. This is an implementation of the processor, not all protocols and transports supported. So if we want to have a full c_glib implementation we should break into smaller issues and give them to different developers.

The current implementation is tested on the basic protocols and transports with some languages (not all) if something is broken, then another issue has to be raised to implement that part on the right way. I cannot afford do the full stack implementation for c_glib in one shot.

gadLinux commented:
In fact @jeking3 Some of the failures are not related to this issue. So please don't ask me to fix them as part of this issue. One of them is caused in the Java Server because the exception management. Maybe caused by the c_glib client. But I'm NOT implementing the client on this issue but the server.


For example:


     [java] 2017-11-13 18:35:41,759 ERROR thrift.ProcessFunction (ProcessFunction.java:process(38)) - Internal error processing testException
     [java] org.apache.thrift.TException: TException
     [java] 	at org.apache.thrift.server.ServerTestBase$TestHandler.testException(ServerTestBase.java:237)
     [java] 	at thrift.test.ThriftTest$Processor$testException.getResult(ThriftTest.java:2089)
     [java] 	at thrift.test.ThriftTest$Processor$testException.getResult(ThriftTest.java:2068)
     [java] 	at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:36)
     [java] 	at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
     [java] 	at org.apache.thrift.TMultiplexedProcessor.process(TMultiplexedProcessor.java:134)
     [java] 	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:291)
     [java] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
     [java] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
     [java] 	at java.lang.Thread.run(Thread.java:745)

gadLinux commented:
Hi @jeking3 I think at the end it was part of the test client... What do you think? It worth it merge it?

gadLinux commented:
Hi @jeking3 can you merge please?

jeking3 commented:
I cannot merge anything while the CI tests are failing due to language specific issues.
```
===============================================================================
*** Following 8 failures were unexpected ***:
If it is introduced by you, please fix it before submitting the code.
===============================================================================
server-client:          protocol:         transport:               result:
c_glib-csharp           multi-binary      buffered-ip              failure(timeout)
c_glib-csharp           multic-compact    buffered-ip              failure(timeout)
c_glib-csharp           multic-compact    framed-ip                failure(timeout)
c_glib-csharp           multi-binary      framed-ip                failure(timeout)
c_glib-rs               multic            framed-ip                failure(timeout)
c_glib-rs               multic            buffered-ip              failure(timeout)
c_glib-rs               multi             framed-ip                failure(timeout)
c_glib-rs               multi             buffered-ip              failure(timeout)
```

This needs to be resolved.  Given that c_glib, csharp, and rs are the only three languages other than Java doing multi, and none of the multi tests are known failures, it looks like the c_glib server is misbehaving OR it is proving that the csharp and rs clients are misbehaving.  Do you know which is the case?

jeking3 commented:
I also fail locally running in the docker container (but csharp does not, which is interesting):
```
===============================================================================
*** Following 4 failures were unexpected ***:
If it is introduced by you, please fix it before submitting the code.
===============================================================================
server-client:          protocol:         transport:               result:
c_glib-rs               multi             buffered-ip              failure(timeout)
c_glib-rs               multi             framed-ip                failure(timeout)
c_glib-rs               multic            buffered-ip              failure(timeout)
c_glib-rs               multic            framed-ip                failure(timeout)
===============================================================================
Unexpected failures are logged to test/log/unexpected_failures.log
You can browse results at:
	file:///thrift/src/test/index.html
# If you use Chrome, run:
# 	cd /thrift/src
#	python -m http.server 8001
# then browse:
# 	http://localhost:8001/test/
Full log for each test is here:
	test/log/server_client_protocol_transport_client.log
	test/log/server_client_protocol_transport_server.log
4 failed of 2590 tests in total.
Test execution took 887.2 seconds.
Wed Nov 15 14:08:42 2017
Makefile:1097: recipe for target 'cross-.*' failed
make: *** [cross-.*] Error 1

RET=$?
if [ $RET -ne 0 ]; then
  cat test/log/unexpected_failures.log
fi
TEST: [c_glib-rs_multi_buffered-ip]
*************************** server message ***************************
Wed Nov 15 13:54:58 2017
Executing: /thrift/src/test/c_glib/test_server --lt-debug --protocol=multi --transport=buffered --port=39553
Directory: /thrift/src/test/c_glib
config:delay: 5
config:timeout: 6
===============================================================================
test_server:test_server:93: libtool wrapper (GNU libtool) 2.4.6 Debian-2.4.6-0.1
test_server:test_server:114: newargv[0]: /thrift/src/test/c_glib/.libs/lt-test_server
test_server:test_server:104: newargv[1]: --protocol=multi
test_server:test_server:104: newargv[2]: --transport=buffered
test_server:test_server:104: newargv[3]: --port=39553
Starting "simple" server (buffered/binary:multi) listen on: 39553

(lt-test_server:36882): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: message type invalid for this processor

(lt-test_server:36882): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 1 bytes - Success
** Message: thrift_simple_server_serve: failed to read 4 bytes - Success

Server process is successfully killed.
===============================================================================
Process is killed.
Test execution took 6.2 seconds.
Wed Nov 15 13:55:04 2017

*************************** client message ***************************
Wed Nov 15 13:54:58 2017
Executing: /thrift/src/test/rs/bin/test_client --protocol=multi --transport=buffered --port=39553
Directory: /thrift/src/test/rs/bin
config:delay: 5
config:timeout: 6
===============================================================================
INFO:test_client: connecting to 127.0.0.1:39553 with multi+buffered stack
INFO:test_client: testVoid
INFO:test_client: testString
INFO:test_client: testBool
INFO:test_client: testBool
INFO:test_client: testByte
INFO:test_client: testi32
INFO:test_client: testi64
INFO:test_client: testDouble
INFO:test_client: testTypedef
INFO:test_client: testEnum
INFO:test_client: testBinary
INFO:test_client: testStruct
INFO:test_client: testNest
INFO:test_client: SecondService blahBlah
===============================================================================
Return code: -9
Test execution took 6.0 seconds.
Wed Nov 15 13:55:04 2017

**********************************************************************

TEST: [c_glib-rs_multi_framed-ip]
*************************** server message ***************************
Wed Nov 15 13:54:58 2017
Executing: /thrift/src/test/c_glib/test_server --lt-debug --protocol=multi --transport=framed --port=44309
Directory: /thrift/src/test/c_glib
config:delay: 5
config:timeout: 6
===============================================================================
test_server:test_server:93: libtool wrapper (GNU libtool) 2.4.6 Debian-2.4.6-0.1
test_server:test_server:114: newargv[0]: /thrift/src/test/c_glib/.libs/lt-test_server
test_server:test_server:104: newargv[1]: --protocol=multi
test_server:test_server:104: newargv[2]: --transport=framed
test_server:test_server:104: newargv[3]: --port=44309
Starting "simple" server (framed/binary:multi) listen on: 44309

(lt-test_server:36899): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: message type invalid for this processor

(lt-test_server:36899): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Success
** Message: thrift_simple_server_serve: failed to read 4 bytes - Success

Server process is successfully killed.
===============================================================================
Process is killed.
Test execution took 6.2 seconds.
Wed Nov 15 13:55:04 2017

*************************** client message ***************************
Wed Nov 15 13:54:58 2017
Executing: /thrift/src/test/rs/bin/test_client --protocol=multi --transport=framed --port=44309
Directory: /thrift/src/test/rs/bin
config:delay: 5
config:timeout: 6
===============================================================================
INFO:test_client: connecting to 127.0.0.1:44309 with multi+framed stack
INFO:test_client: testVoid
INFO:test_client: testString
INFO:test_client: testBool
INFO:test_client: testBool
INFO:test_client: testByte
INFO:test_client: testi32
INFO:test_client: testi64
INFO:test_client: testDouble
INFO:test_client: testTypedef
INFO:test_client: testEnum
INFO:test_client: testBinary
INFO:test_client: testStruct
INFO:test_client: testNest
INFO:test_client: SecondService blahBlah
===============================================================================
Return code: -9
Test execution took 6.0 seconds.
Wed Nov 15 13:55:04 2017

**********************************************************************

TEST: [c_glib-rs_multic_buffered-ip]
*************************** server message ***************************
Wed Nov 15 13:55:00 2017
Executing: /thrift/src/test/c_glib/test_server --lt-debug --protocol=multic --transport=buffered --port=37943
Directory: /thrift/src/test/c_glib
config:delay: 5
config:timeout: 6
===============================================================================
test_server:test_server:93: libtool wrapper (GNU libtool) 2.4.6 Debian-2.4.6-0.1
test_server:test_server:114: newargv[0]: /thrift/src/test/c_glib/.libs/lt-test_server
test_server:test_server:104: newargv[1]: --protocol=multic
test_server:test_server:104: newargv[2]: --transport=buffered
test_server:test_server:104: newargv[3]: --port=37943
Starting "simple" server (buffered/compact:multic) listen on: 37943

(lt-test_server:36956): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: message type invalid for this processor

(lt-test_server:36956): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 1 bytes - Success
** Message: thrift_simple_server_serve: failed to read 1 bytes - Success

Server process is successfully killed.
===============================================================================
Process is killed.
Test execution took 6.2 seconds.
Wed Nov 15 13:55:07 2017

*************************** client message ***************************
Wed Nov 15 13:55:01 2017
Executing: /thrift/src/test/rs/bin/test_client --protocol=multic --transport=buffered --port=37943
Directory: /thrift/src/test/rs/bin
config:delay: 5
config:timeout: 6
===============================================================================
INFO:test_client: connecting to 127.0.0.1:37943 with multic+buffered stack
INFO:test_client: testVoid
INFO:test_client: testString
INFO:test_client: testBool
INFO:test_client: testBool
INFO:test_client: testByte
INFO:test_client: testi32
INFO:test_client: testi64
INFO:test_client: testDouble
INFO:test_client: testTypedef
INFO:test_client: testEnum
INFO:test_client: testBinary
INFO:test_client: testStruct
INFO:test_client: testNest
INFO:test_client: SecondService blahBlah
===============================================================================
Return code: -9
Test execution took 6.0 seconds.
Wed Nov 15 13:55:07 2017

**********************************************************************

TEST: [c_glib-rs_multic_framed-ip]
*************************** server message ***************************
Wed Nov 15 13:55:01 2017
Executing: /thrift/src/test/c_glib/test_server --lt-debug --protocol=multic --transport=framed --port=42147
Directory: /thrift/src/test/c_glib
config:delay: 5
config:timeout: 6
===============================================================================
test_server:test_server:93: libtool wrapper (GNU libtool) 2.4.6 Debian-2.4.6-0.1
test_server:test_server:114: newargv[0]: /thrift/src/test/c_glib/.libs/lt-test_server
test_server:test_server:104: newargv[1]: --protocol=multic
test_server:test_server:104: newargv[2]: --transport=framed
test_server:test_server:104: newargv[3]: --port=42147
Starting "simple" server (framed/compact:multic) listen on: 42147

(lt-test_server:36986): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: message type invalid for this processor

(lt-test_server:36986): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: failed to read 4 bytes - Success
** Message: thrift_simple_server_serve: failed to read 4 bytes - Success

Server process is successfully killed.
===============================================================================
Process is killed.
Test execution took 6.2 seconds.
Wed Nov 15 13:55:07 2017

*************************** client message ***************************
Wed Nov 15 13:55:01 2017
Executing: /thrift/src/test/rs/bin/test_client --protocol=multic --transport=framed --port=42147
Directory: /thrift/src/test/rs/bin
config:delay: 5
config:timeout: 6
===============================================================================
INFO:test_client: connecting to 127.0.0.1:42147 with multic+framed stack
INFO:test_client: testVoid
INFO:test_client: testString
INFO:test_client: testBool
INFO:test_client: testBool
INFO:test_client: testByte
INFO:test_client: testi32
INFO:test_client: testi64
INFO:test_client: testDouble
INFO:test_client: testTypedef
INFO:test_client: testEnum
INFO:test_client: testBinary
INFO:test_client: testStruct
INFO:test_client: testNest
INFO:test_client: SecondService blahBlah
===============================================================================
Return code: -9
Test execution took 6.0 seconds.
Wed Nov 15 13:55:07 2017

**********************************************************************
```

gadLinux commented:
Can it be because the SSL server. That's not implemented on c_glib?

jeking3 commented:
The tests that failed do not use ssl.

gadLinux commented:
But I don't know about the languages that are failing. This is why I told you to open other issue and merge this.

jeking3 commented:
I cannot merge in a change that will knowingly break CI builds.

The tests that are failing need to be fixed (preferably) or disabled using the "test.py -U merge" I (hopefully) documented earlier.  Once the failing tests are disabled and the build passes we can merge.  We have a backlog of "known failing" test combinations that need attention and are not represented in the Jira backlog.

jeking3 commented:
Under THRIFT-2013 I am adding multiplexed cpp testing to the suite.  Will see how it interacts.  If only "rs" client is misbehaving, we can disable.

gadLinux commented:
Just curious that The error messages I get are different. In fact they got blocked on the RS (Rust?) client, at least it seems to.

gadLinux commented:
I tried to debug the Rust thing. But it seems is not receiving the response from c_glib testNest in my case. I debugged the server and the response is sent, and flushed. So I don't catch why RS is not receiving it. In fact the client stalls waiting a response. I'm more convinced now that it should be looked by a Rust develper. Enable c++ test if you want and I will take a look if they fail against c_glib.

jeking3 commented:
I'm going to carry your changes into another PR that has cpp client and server multiplexed and java multiplexed testing added, and fixes some other bugs I found along the way.  Once it builds clean locally I will submit a PR that contains THRIFT-4329 entirely, plus those changes.  I have set the c_glib-rs multi tests as known failures for now.  It is behaving well against cpp and java.

jeking3 commented:
@gadLinux you can close this PR if you want; I will carry it forward in a combined PR for our stuff: https://github.com/apache/thrift/pull/1414

gadLinux commented:
Hi @jeking3 it was a hard feature to merge... Hope we will do finer next time. Thank you for your continued support.
